src/plugins/mpath.c

```diff
@@ -156,360 +184,295 @@ gboolean bd_mpath_is_tech_avail (BDMpathTech tech, guint64 mode, GError **error)
-    g_free (path);
-    if (!link) {
-        g_prefix_error (error, "Failed to determine device name for '%s'",
-                        major_minor);
-        return NULL;
-    }
-
-    /* 'link' should be something like "../sda" */
-    /* get the last '/' */
-    ret = strrchr (link, '/');
-    if (!ret) {
-        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_INVAL,
-                     "Failed to determine device name for '%s'",
-                     major_minor);
-        g_free (link);
-        return NULL;
+    _dmmp_ctx_setup (ctx, err_msg_buff);
+    if (ctx == NULL) {
+        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_NO_MEMORY,
+                     "No enough memory to create dmmp context");
+        goto out;
     }
-    /* move right after the last '/' */
-    ret++;
 
-    /* create a new copy and free the whole link path */
-    ret = g_strdup (ret);
-    g_free (link);
+    dmmp_rc = dmmp_mpath_array_get (ctx, &mps, &mp_count);
 
-    return ret;
-}
+    if (dmmp_rc != DMMP_OK)
+        goto out;
 
-static gboolean map_is_multipath (const gchar *map_name, GError **error) {
-    struct dm_task *task = NULL;
-    struct dm_info info;
-    guint64 start = 0;
-    guint64 length = 0;
-    gchar *type = NULL;
-    gchar *params = NULL;
-    gboolean ret = FALSE;
-
-    if (geteuid () != 0) {
-        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_NOT_ROOT,
-                     "Not running as root, cannot query DM maps");
-        return FALSE;
+    if (mp_count == 0) {
+        rc = TRUE;
+        goto out;
     }
 
-    task = dm_task_create (DM_DEVICE_STATUS);
-    if (!task) {
-        g_warning ("Failed to create DM task");
-        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_DM_ERROR,
-                     "Failed to create DM task");
-        return FALSE;
+    for (i = 0; i < mp_count; ++i) {
+        name = dmmp_mpath_name_get (mps[i]);
+        dmmp_flush_mpath (ctx, name);
     }
+    dmmp_rc = dmmp_mpath_array_get (ctx, &mps, &mp_count);
 
-    if (dm_task_set_name (task, map_name) == 0) {
-        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_DM_ERROR,
-                     "Failed to create DM task");
-        dm_task_destroy (task);
-        return FALSE;
-    }
+    if (dmmp_rc != DMMP_OK)
+        goto out;
 
-    if (dm_task_run (task) == 0) {
-        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_DM_ERROR,
-                     "Failed to run DM task");
-        dm_task_destroy (task);
-        return FALSE;
-    }
+    if (mp_count == 0)
+        rc = TRUE;
 
-    if (dm_task_get_info (task, &info) == 0) {
-        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_DM_ERROR,
-                     "Failed to get task info");
-        dm_task_destroy (task);
-        return FALSE;
+out:
+    if (dmmp_rc != DMMP_OK) {
+        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_FLUSH,
+                     "%d: %s", dmmp_rc, _dmmp_last_err_msg(ctx));
```

Comment:
This should provide some more descriptive error messages based on what (step) exactly failed.

Reply:
Will fix this by removing `goto` statement.